pkg/resource/deploy/deployment_executor.go

```diff
@@ -242,15 +227,23 @@ func (ex *deploymentExecutor) Execute(callerCtx context.Context, opts Options, p
 				}
 
 				if event.Event == nil {
-					res := ex.performDeletes(ctx, updateTargetsOpt, destroyTargetsOpt)
+					// Check targets before performDeletes mutates the initial Snapshot.
+					targetErr := ex.checkTargets(opts.Targets)
+
+					res := ex.performDeletes(ctx, opts.Targets)
```

Comment:
This seems related to the side effect that I've commented:
- https://github.com/pulumi/pulumi/issues/8164